### PR TITLE
Make `state.narrows` an Immutable.Map.

### PR DESCRIPTION
**Pull request by chrisbobbe**
An instance of #3949 and #3950, this is planned followup to #4047. See https://github.com/zulip/zulip-mobile/pull/4047#issuecomment-616877699.

~~Still getting all these tests type-checked, but thought I'd push my work today.~~ 🙂

Done! Well, with a few things to note:

- `src/message/__tests__/fetchActions-test.js` isn't type-checked and made to use example data in this PR. Those are done in my current branch for #4171, which is open. No need to cherry-pick that work, I think; that'll just make things more confusing. I don't think this has a strong influence on which PR we should aim to merge first: I just want to generally be sure that the test files I touch will be type-checked at some point.

- I did cherry-pick a few commits from my branch for #4171, only in support of type-checking other test files for the first time.

- ~~I also cherry-picked one commit from #3946, which is also open.~~ #3946 was recently merged, so no need—just rebased across this. 🙂 

I've called out all of these cherry-picks with a note at the bottom of the commit message.

### CODE REVIEW

**Comment by chrisbobbe** [comment]
Just resolved some conflicts; this is ready for review.

File: src/chat/__tests__/narrowsSelectors-test.js [hunk 1]
```diff
@@ -120,11 +114,9 @@ describe('getMessagesForNarrow', () => {
       ],
     });
 
-    const anchor = getMessagesForNarrow(state, privateNarrow('john@example.com'));
-
-    const expectedState = deepFreeze([{ id: 123 }]);
```
Comment (gnprice) [inline src/chat/__tests__/narrowsSelectors-test.js]: Much better -- thanks for cleaning these up :slightly_smiling_face: 

I think with these names you're looking at the results of two different cases of someone copy-pasting without thinking it through (or perhaps not understanding.) The `expectedState` is, as you say, a reasonable name in many reducer tests.

The `anchor`... I suspect comes from the last few test cases below, testing things like `getFirstMessageId`. There, the result is at least plausibly an "anchor" -- it's a message ID that we're going to use to identify a spot in the message stream. And then those got copied elsewhere, by a contributor who didn't understand what it meant and in a period where we weren't doing meaningful code review.

OK, and having said that I was curious to find out. :slightly_smiling_face: Indeed -- and another layer is that those tests for `getFirstMessageId` etc. were originally for a function called... `getAnchor`. a89d897c2 changed that. And before then, 8dd208120 propagated the `anchor` name to these other test cases.

File: src/pm-conversations/__tests__/pmConversationsSelectors-test.js [hunk 2]
```diff
@@ -23,103 +27,117 @@ describe('getRecentConversations', () => {
...
+      display_recipient: [eg.displayRecipientFromUser(eg.selfUser)],
+    });
+
+    const meJohnAndMarkPm = eg.pmMessage({
+      id: 0,
+      display_recipient: [
+        eg.displayRecipientFromUser(eg.selfUser),
+        eg.displayRecipientFromUser(userMark),
+        eg.displayRecipientFromUser(userJohn),
+      ],
```
Comment (gnprice) [inline src/pm-conversations/__tests__/pmConversationsSelectors-test.js:67]: These are probably a sign that it'd be helpful to refactor `eg.displayRecipientFromUser` to be called, say, `eg.displayRecipientFromUsers` -- and to take `User[]` and return `PmRecipientUser[]`, and be just a `map` around what it currently is.

(Maybe even both versions should exist.)

**Comment by gnprice** [comment]
... Err, I guess that came out slightly wrong, in that this PR is marked as merged but with that headline final commit dropped, oops. Please go ahead and open a new PR with (a rebased version of) 1a80243f6.

The set of commits merged to master were exactly what I intended, though :slightly_smiling_face:

**Comment by chrisbobbe** [comment]
OK, opened as #4201!

**Pull request merged**